File: ol_control_receive.go
//@note: This is based on an outdated version of ol_control_receive.py

package main

import (
	"encoding/binary"
	"fmt"
	"log"
	"net"
	"os"
	"os/signal"
	"syscall"
	"time"
)

const (
	address       = "0.0.0.0:12345"
	flushInterval = 1
)

func main() {
	addr, err := net.ResolveUDPAddr("udp", address)
	if err != nil {
		log.Fatalf("Failed to resolve address: %v", err)
	}

	conn, err := net.ListenUDP("udp", addr)
	if err != nil {
		log.Fatalf("Failed to listen: %v", err)
	}
	defer conn.Close()

	file, err := os.OpenFile("ol_data_log.txt", os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
	if err != nil {
		log.Fatalf("Failed to open file: %v", err)
	}
	defer file.Close()

	packetCounter := 0

	sigChan := make(chan os.Signal, 1)
	signal.Notify(sigChan, os.Interrupt, syscall.SIGTERM)

	for {
		select {
		case <-sigChan:
			fmt.Println("\n[!] Keyboard interrupt received\n[*] Closing the program...")
			os.Exit(0)

		default:
			err := conn.SetReadDeadline(time.Now().Add(1 * time.Second))
			if err != nil {
				log.Fatalf("Failed to set deadline: %v", err)
			}

			buf := make([]byte, 1024)
			n, srcAddr, err := conn.ReadFromUDP(buf)

			if err != nil {
				if netErr, ok := err.(net.Error); ok && netErr.Timeout() {
					continue
				}
				log.Printf("Error reading from UDP: %v", err)
				continue
			}

			packetCounter++

			srcIP := buf[0:4]
			dstIP := buf[4:8]
			payloadLength := binary.BigEndian.Uint16(buf[8:10])
			sequenceNumber := binary.BigEndian.Uint16(buf[10:12])

			payload := buf[12:n]

			fmt.Printf("Received packet from %s\n", srcAddr)
			fmt.Printf("Source Address: %s\n", net.IP(srcIP))
			fmt.Printf("Destination Address: %s\n", net.IP(dstIP))
			fmt.Printf("Payload Length: %d\n", payloadLength)
			fmt.Printf("Sequence Number: %d\n", sequenceNumber)
			fmt.Printf("Payload Data: %x\n", payload)

			file.WriteString(fmt.Sprintf("Received packet from %s\n", srcAddr))
			file.WriteString(fmt.Sprintf("Source Address: %s\n", net.IP(srcIP)))
			file.WriteString(fmt.Sprintf("Destination Address: %s\n", net.IP(dstIP)))
			file.WriteString(fmt.Sprintf("Payload Length: %d\n", payloadLength))
			file.WriteString(fmt.Sprintf("Sequence Number: %d\n", sequenceNumber))
			file.WriteString(fmt.Sprintf("Payload Data: %x\n\n", payload))

			if packetCounter%flushInterval == 0 {
				file.Sync()
				fmt.Println("[*] Data written to file.")
			}
		}
	}
}

File: archive/ol_control_receive.py
# Archived since compatibility issues with WinXP and Python 3.4
import socket
import struct
import sys
import select
import time

# Define the packet header format
HEADER_FORMAT = '!4s4sBBHI'

# Create a socket and bind to a port to listen for incoming packets
s = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
s.bind(('0.0.0.0', 12345))  # Listen on all interfaces on port 12345

packet_counter = 0
flush_interval = 1

try:
    with open('ol_data_log.txt', 'ab') as f:  # Open the file in binary append mode
        while True:
            # Wait for incoming packets with a timeout of 1 second
            ready, _, _ = select.select([s], [], [], 1.0)

            if ready:
                packet_counter += 1

                # Receive a packet
                packet, address = s.recvfrom(1024)

                # Unpack the header fields from the packet
                source_address, destination_address, protocol_id, payload_length, sequence_number, timestamp = struct.unpack(HEADER_FORMAT, packet[:16])

                # Extract the payload data from the packet
                payload = packet[12:]

                # Print the packet information and payload data
                print(f'Received packet from {address[0]}:{address[1]}')
                print(f'Source Address: {socket.inet_ntoa(source_address)}')
                print(f'Destination Address: {socket.inet_ntoa(destination_address)}')
                print(f'Payload Length: {payload_length}')
                print(f'Sequence Number: {sequence_number}')
                timestamp_str = time.strftime('%Y-%m-%d %H:%M:%S', time.localtime(timestamp))
                print(f'Timestamp: {timestamp_str}')
                print(f'Payload Data: {payload}')

                # Write the packet information and payload data to the file
                f.write(f'Received packet from {address[0]}:{address[1]}\n'.encode())
                f.write(f'Source Address: {socket.inet_ntoa(source_address)}\n'.encode())
                f.write(f'Destination Address: {socket.inet_ntoa(destination_address)}\n'.encode())
                f.write(f'Payload Length: {payload_length}\n'.encode())
                f.write(f'Sequence Number: {sequence_number}\n'.encode())
                f.write(f'Timestamp: {timestamp_str}\n'.encode())
                f.write(f'Payload Data: {payload}\n\n'.encode())

                if packet_counter % flush_interval == 0:
                    f.flush()
                    print('[*] Data written to file.')

except KeyboardInterrupt:
    print('[!] Keyboard interrupt received\n[*] Closing the program...')
    s.close()
    sys.exit(0)